components/compliance-service/reporting/relaxting/suggestions.go

```diff
@@ -502,6 +502,170 @@ func (backend ES2Backend) getControlSuggestions(client *elastic.Client, typePara
+
+	hit := elastic.NewInnerHit().Size(20)
+	boolQuery = boolQuery.Must(elastic.NewNestedQuery("profiles.controls", nestedBoolQuery).InnerHit(hit))
+
+	searchSource := elastic.NewSearchSource().
+		Query(boolQuery).
+		FetchSource(false).
+		Size(size * 50) // Multiplying size to ensure that same profile with multiple versions is not limiting our suggestions to a lower number
+		// ^ Because we can't sort by max_score of the inner hits: https://discuss.elastic.co/t/nested-objects-hits-inner-hits-and-sorting/32565
+
+	source, err := searchSource.Source()
+	if err != nil {
+		return nil, errors.Wrap(err, "getControlTagsSuggestions unable to get Source")
+	}
+
+	LogQueryPartMin(esIndex, source, "getControlTagsSuggestions query")
+
+	searchResult, err := client.Search().
+		SearchSource(searchSource).
+		Index(esIndex).
+		FilterPath(
+			"took",
+			"hits.total",
+			"hits.hits._id",
+			"hits.hits._score",
+			"hits.hits.inner_hits").
+		Do(context.Background())
+
+	if err != nil {
+		logrus.Error("getControlTagsSuggestions search failed")
+		return nil, err
+	}
+
+	logrus.Debugf("Search query took %d milliseconds\n", searchResult.TookInMillis)
+
+	type ControlSourceTags struct {
+		ID         string                            `json:"id"`
+		StringTags []ESInSpecReportControlStringTags `json:"string_tags"`
+	}
+	type ControlTagsScore struct {
+		Score      float32
+		StringTags []ESInSpecReportControlStringTags
+	}
+
+	// Using this array to aggregate the unsorted suggestions from the response
+	foundControlTags := make([]ControlTagsScore, 0)
+	if searchResult != nil {
+		for _, hit := range searchResult.Hits.Hits {
+			if hit == nil {
+				continue
+			}
+			for _, inner_hit := range hit.InnerHits {
+				if inner_hit == nil {
+					continue
+				}
+				for _, hit2 := range inner_hit.Hits.Hits {
+					if hit2.Source == nil {
+						continue
+					}
+					var c ControlSourceTags
+					err := json.Unmarshal(*hit2.Source, &c)
+					if err == nil && c.ID != "" {
+						foundControlTags = append(foundControlTags, ControlTagsScore{
+							float32(*hit2.Score),
+							c.StringTags,
+						})
+					}
+				}
+			}
+		}
+	}
+	sort.Slice(foundControlTags, func(i, j int) bool {
+		return foundControlTags[i].Score > foundControlTags[j].Score
+	})
+
+	// Map the found tag suggestions to remove duplications
+	scoredTagSuggs := make(map[string]*float32)
+	for _, item := range foundControlTags {
+		matches := make([]string, 0)
+		// If the suggestion is for control tag key, we need to find the one that
+		// was the closest match from all the tag keys the control might have
+		if typeParam == "control_tag_key" {
+			for _, controlTag := range item.StringTags {
+				matches = append(matches, controlTag.Key)
+			}
+		}
+
+		// If the suggestion is for control tag values, we need to find the tag that matches the
+		// tag key and then find the closest match from all the values of that tag
+		if typeParam == "control_tag_value" {
```

Comment:
Any value in doing this append in the pass we are doing in 612-614 above? I think the conditions are the same, so we could move 621-623 into the same loop unless I've misread.

Reply:
Thank you Steven.
Just to make sure I get it right, you suggest refactoring this code:
```go
		// If the suggestion is for control tag key, we need to find the one that
		// was the closest match from all the tag keys the control might have
		if typeParam == "control_tag_key" {
			for _, controlTag := range item.StringTags {
				matches = append(matches, controlTag.Key)
			}
		}

		// If the suggestion is for control tag values, we need to find the tag that matches the
		// tag key and then find the closest match from all the values of that tag
		if typeParam == "control_tag_value" {
			for _, controlTag := range item.StringTags {
				if controlTag.Key == controlTagFilterKey {
					matches = append(matches, controlTag.Values...)
				}
			}
		}
```

to this?

```go
		for _, controlTag := range item.StringTags {
			// If the suggestion is for control tag key, we need to find the one that
			// was the closest match from all the tag keys the control might have
			// If the suggestion is for control tag values, we need to find the tag that matches the
			// tag key and then find the closest match from all the values of that tag
			if typeParam == "control_tag_key" {
				matches = append(matches, controlTag.Key)
			} else if typeParam == "control_tag_value" && controlTag.Key == controlTagFilterKey {
				matches = append(matches, controlTag.Values...)
			}
		}
```

Reply:
Ha! I misread this when reviewing.  For some reason I didn't see that the typeParam comparison here was different than above.  Looks good as-is.

Reply:
Thank you, I split the code like that to make it easier to read.